docs/rfcs/confmap-provider-configuration.md

```diff
@@ -0,0 +1,184 @@
+interfaces are still able to change before the `confmap` module is declared
+stable, but avoiding breaking changes in the API would be preferable.
+
+## Desired state
+
+We would like a mechanism to configure Providers that meets the following
+requirements:
+
+1. Broadly applicable to all current upstream providers.
+2. Can be used as a suggestion for configuring custom providers.
+3. Is consistent between Providers.
+4. Is configurable per config URI.
+
+## Solutions
+
+Providers are invoked through passing `--config` flags to the Collector binary
+with a scheme and URI to obtain. A single instance of each Provider is then
+tasked with retrieving from config from all URIs passed to the Collector for the
+scheme the provider is registered to handle.
+
+We have the following places where it may make sense to hook into options for
+Providers:
+
+1. Parts of the URI we are requesting.
+2. Part of the string passed to `--config`.
+3. Separate flags to configure Providers per config URI.
+4. Use a separate config file that specifies config sources inside a map
+   structure.
+5. Extend the Collector's config schema to support specifying additional places
+   to obtain configuration
+
+### Configure options inside the URI
+
+[RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3), which
+specifies the format of a URI, specifies the different parts of a URI and
+suggests two places where we could pass options to Providers: queries and
+fragments.
+
+#### Queries
+
+Advantages:
+
+- Explicitly intended to  specify non-hierarchical data in a URI.
+- Often used for this purpose.
+- Fits into existing config URIs.
+
+Disadvantages:
+
+- Completely consuming the parameters would mean we can't pass them to config
+  backends. The frequent use of query parameters in server endpoints would force
+  us to limit the types of requests we can make.
+- Partially consuming parameters would still cause conflicts if a Provider
+  parameter and config backend parameter share the same name. Additionally,
+  mixing the two would provide suboptimal UX.
+- Only allows easily specifying key-value pairs.
+
+#### Fragments
+
+We could specify options in a query parameter-encoded string placed into the URI
+fragment.
+
+Advantages:
+
+- Not likely to be used by config backends for any of our supported protocols,
+  so has a low chance of conflict.
+- Fits into existing config URIs.
+
+Disadvantages:
+
+- Even if fragments are likely not useful to backends, we are still preventing
+  their use in upstream Providers.
+- Doesn't conform to the spirit of how fragments should be used according to RFC
+  3986.
+- Only allows easily specifying key-value pairs.
+
+We could likely partially circumvent the key-value pair limitation by
+recursively calling confmap Providers to resolve files, env vars, HTTP URLs,
+etc. For example:
+
+```text
+https://config.com/config#refresh-interval=env:REFRESH_INTERVAL&headers=file:headers.yaml
+```
```

Comment:
You can do this since #7504 with the following syntax:
```text
https://config.com/config#refresh-interval=${env:REFRESH_INTERVAL}&headers=${file:headers.yaml}
```

At least, within a configuration file, I am not sure about CLI

Reply:
I think the braces syntax only applies to URIs in files and doesn't work in the CLI. The reason I simplified the syntax a little was to avoid requiring users to escape characters that a shell may consume like `$`. On the other hand, we need some obvious way to indicate that these are URIs for extra config and not values in themselves. I'm not sure the way I have it written right now will suffice for that.

We also need to consider the way the substitution works for these options. If we have the confmap Resolver do resolution for embedded config URIs, `headers.yaml` would need to be marshaled into a single line, such that this:

```text
https://config.com/config#refresh-interval=${env:REFRESH_INTERVAL}&headers=${file:headers.yaml}
```

would resolve to this:

```text
https://config.com/config#refresh-interval=10s&headers=['Api-Token':'0xdeadbeef' 'Accept-Encoding':'text/yaml']
```

The Resolver would need to know when to "stringify" files like this vs. when to substitute the direct contents of a file. Or we just require users to write the stringified version, which simplifies things for us at the cost of worse UX since in my opinion this representation would be unpleasant to work with.

My assumption was that we implement custom unmarshaling logic for the URI fragment options that takes all of this into account. So `${file:headers.yaml}` is unmarshaled directly into a map that is assigned to a `Headers` field on a Provider's config struct.

Reply:
> The reason I simplified the syntax a little was to avoid requiring users to escape characters that a shell may consume like $.

I am a bit hesitant about dropping `$` both for consistency and because `#refresh-interval={env:REFRESH_INTERVAL}` is (I think?) a valid part of an URI, so I wouldn't want to prevent users from writing that.

> The Resolver would need to know when to "stringify" files like this vs. when to substitute the direct contents of a file. 

My expectation is that this would work the same way it does if you use an `${...}` syntax in the middle of a string in the actual config.

Reply:
>I am a bit hesitant about dropping $ both for consistency and because #refresh-interval={env:REFRESH_INTERVAL} is (I think?) a valid part of an URI, so I wouldn't want to prevent users from writing that.

It would be a valid part of the URI, but the assumption here is that we reserve the fragment portion of URIs for our own use in upstream confmap components. Custom components would be free to use this portion of the URI as they wish, including forwarding it to a backend. I'm fine either way with whether recursive resolution requires `$`.

>My expectation is that this would work the same way it does if you use an ${...} syntax in the middle of a string in the actual config.

I'm alright with that, that's probably the least surprising behavior here.